easybuild/tools/filetools.py

```diff
@@ -1265,14 +1265,11 @@ def verify_checksum(path, checksums):
 
     for checksum in checksums:
         if isinstance(checksum, dict):
-            if filename in checksum:
+            try:
                 # Set this to a string-type checksum
                 checksum = checksum[filename]
-            elif build_option('enforce_checksums'):
-                raise EasyBuildError("Missing checksum for %s", filename)
-            else:
-                # Set to None and allow to fail elsewhere
-                checksum = None
+            except KeyError:
+                raise EasyBuildError("Missing checksum for %s in %s", filename, checksum)
```

Comment:
I wanted to keep the current behavior that this situation is an error and just provide a meaningful error message.

I drafted https://github.com/easybuilders/easybuild-docs/pull/104 where there are some open question on how checksums should be handled which is likely a breaking change (or a handful of breaking changes)

IMO a dict with a missing filename/key is an error: Likely the dict was meant for arch-specific files&checksums but a new arch is used which wasn't tested yet. Or a spelling error happened making it always fail to verify checksums even if it should.    
That's what I suggested in the draft mentioned above: A dict, if present, needs to list each possible filename even without `enforce-checksums`

Reply:
BTW: With a [recent issue](https://github.com/easybuilders/easybuild-easyconfigs/pull/18491#issuecomment-1670182732) I noticed that `enforce-checksums` might be to greedy: We **need** a `None` value if the archive (e.g. Git checkout) isn't checksummable. But we want to check that all other checksums are present.

Now we don't (usually) have lists anymore but lists of dicts of filename-to-checksum. Now if `enforce-checksums` triggers for every explicit `None` but a missing filename in a dict doesn't trigger an error if it isn't set we have no way to check e.g. the (old) PyTorch ECs reliably.

Hence I think the approach I suggested makes sense.

Reply:
Thanks for clarifying, I agree that this is at least a good step forward to avoid totally meaningless errors.

If you noticed any other problems, please make sure there's an issue open for it (or if not, open one)